src/components/IdleTimeoutModal.js

```diff
@@ -0,0 +1,105 @@
+import _ from 'lodash/fp'
+import { useEffect, useState } from 'react'
+import { div, h, iframe } from 'react-hyperscript-helpers'
+import ButtonBar from 'src/components/ButtonBar'
+import Modal from 'src/components/Modal'
+import { getUser } from 'src/libs/auth'
+import colors from 'src/libs/colors'
+import { authStore, lastActiveTimeStore } from 'src/libs/state'
+import * as Utils from 'src/libs/utils'
+
+
+const displayRemainingTime = remainingSeconds => {
+  return _.join(':', [
+    `${Math.floor(remainingSeconds / 60).toString().padStart(2, '0')}`,
+    `${Math.ceil(remainingSeconds % 60).toString().padStart(2, '0')}`
+  ])
+}
+
+const setLastActive = lastActive => lastActiveTimeStore.update(existing => ({ ...existing, [getUser().id]: lastActive }))
+
+const IdleTimeoutModal = ({
+  timeout = Utils.hhmmssToMs({ mm: 15 }),
+  countdownStart = Utils.hhmmssToMs({ mm: 3 }), emailDomain = ''
+}) => {
+  const { isSignedIn, profile: { email }, user: { id } } = Utils.useStore(authStore)
+  const isClinicalDomain = email && emailDomain ? email.includes(`@${emailDomain}`) : false
+  const activeTimeStore = Utils.useStore(lastActiveTimeStore)
+
+  const currentTime = Date.now()
+  const lastActive = activeTimeStore[id] === 'expired' ? null : activeTimeStore[id]
+  const lastActiveTime = lastActive ? parseInt(lastActive, 10) : currentTime
+  const timedOut = currentTime > lastActiveTime + timeout
+
+  useEffect(() => { timedOut && !isSignedIn && setLastActive('expired') }, [isSignedIn, timedOut])
+
+  return Utils.cond(
+    [isSignedIn && isClinicalDomain, h(InactivityTimer, { id, timeout, countdownStart })],
+    [activeTimeStore[id] === 'expired' && !isSignedIn, () => h(Modal, {
+      title: 'Session Expired',
+      showCancel: false,
+      onDismiss: () => setLastActive(),
+      onOk: () => setLastActive()
+    }, ['Your session has expired to maintain security and protect clinical data'])],
+    null)
+}
+
+const InactivityTimer = ({ id, timeout, countdownStart }) => {
+  const activeTimeStore = Utils.useStore(lastActiveTimeStore)
+  const [logoutRequested, setLogoutRequested] = useState()
+  const [currentTime, setDelay] = Utils.useCurrentTime()
+
+  const lastActive = activeTimeStore[id] === 'expired' ? Date.now() : activeTimeStore[id]
+  const lastActiveTime = lastActive ? parseInt(lastActive, 10) : Date.now()
+  const timeoutTime = lastActiveTime + timeout
+  const timedOut = currentTime > timeoutTime
```

Comment:
These are ever so slightly different than the calculations in `IdleTimeoutModal`. Is that intentional/necessary? This stuff is always a pain to think through. It would be great to have a function that you pass a few facts and get back a set of derived properties.